projects/limtoc/main.cc

```diff
@@ -0,0 +1,153 @@
+/*
+    ChibiOS - Copyright (C) 2006..2015 Giovanni Di Sirio
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include "ch.h"
+#include "hal.h"
+
+#include "analog.h"
+#include "encoder.h"
+
+#include "gitsha1.h"
+#include "blink.h"
+#include "usbconfig.h"
+#include "saconfig.h"
+#include "utility.h"
+#include "printf.h"
+
+#include "parameters.h"
+
+#include <cstdarg>
+
+namespace {
+    constexpr systime_t dt = MS2ST(1); // milliseconds converted to system ticks
+    constexpr float k = 3.14f; // N-m/rad, desired spring stiffness
+    constexpr float m_star = 1.0f; // kg-m^2, desired full assembly virtual inertia
+    constexpr float m_upper = 0.0413f; // kg-m^2, experimentally determined upper assembly inertia
+    constexpr float m_lower = 0.0415f; // kg-m^2, experimentally determined lower assembly inertia
+    constexpr float m_star_lower = m_star - m_upper; // kg-m^2, desired lower assembly virtual inertia
+    constexpr float K = m_lower/m_star_lower - 1.0f; // torque reference feedback gain
+
+    // sensors
+    Analog analog;
+    Encoder encoder_steer(sa::RLS_ROLIN_ENC, sa::RLS_ROLIN_ENC_INDEX_CFG);
+
+    dacsample_t set_handlebar_reference(float reference) {
+        constexpr float MAX_REF_VALUE = sa::MAX_KOLLMORGEN_TORQUE;
+        const float saturated_reference = util::clamp(reference, -MAX_REF_VALUE, MAX_REF_VALUE);
+        const dacsample_t aout = saturated_reference/MAX_REF_VALUE*sa::DAC_HALF_RANGE + sa::DAC_HALF_RANGE;
+        dacPutChannelX(sa::KOLLM_DAC, 0, aout); // TODO: don't hardcode zero but find the DAC channel constant
```

Comment:
This is implemented correctly here: https://github.com/oliverlee/phobos/commit/258cb7027540e55514bcefb9ba71ad5539b6a191

I'll fix it when I rebase off master.